### PR TITLE
Transaction History

### PR DESCRIPTION
**Pull request by siy**

### CODE REVIEW

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/MessageEntry.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+		return new MessageEntry(message, encryptionScheme);
+	}
+
+	public static Optional<MessageEntry> fromPlainString(String message) {
+		return Optional.ofNullable(message).map(text -> create(text, PLAIN));
+	}
+
+	public JSONObject asJson() {
+		return jsonObject().put("msg", message).put("encryptionScheme", encryptionScheme);
```
Comment (Sajjon) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/MessageEntry.java:52]: reminder: we should remove `encryptionScheme` and change:

from OBJECT
```
{
  "msg": STRING,
  "encryptionScheme"
}
```

TO STRING
`msg: STRING` // max 255 bytes, i.e. max 510 hex chars.

File: radixdlt-core/radixdlt/src/test/java/org/radix/api/jsonrpc/HighLevelApiHandlerTest.java [hunk 2]
```diff
@@ -107,6 +116,43 @@ public void testTokenInfo() {
...
+		assertNotNull(response);
+
+		var result = response.getJSONObject("result");
+
+		assertTrue(result.has("cursor"));
+		assertTrue(result.has("transactions"));
+		var transactions = result.getJSONArray("transactions");
+		assertEquals(1, transactions.length());
+
+		var singleTransaction = transactions.getJSONObject(0);
```
Comment (Sajjon) [inline radixdlt-core/radixdlt/src/test/java/org/radix/api/jsonrpc/HighLevelApiHandlerTest.java:145]: Assert `"sentAt"` date field and `"txID"` field? :)

**Review by Sajjon — COMMENTED** [review]
Big PR :D I did not read it carefully and mostly looked at tests for transactionHistory, but looked about right! I think I only saw one test containing one one action, of type `Other`, I would maybe have included one test for tx with single `TransferTokens` action, one test for single `StakeTokens` and one for single `UnstakeTokens`  and maybe one test for a tx with multiple actions, but that is up to you and network team :)

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/MessageEntry.java (hunk 1 above)
Comment (siy) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/MessageEntry.java:52]: This change is planned for next PR. I'm going to implement it as the byte array.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/ClientApiUtils.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	}
+
+	private static Tuple2<List<REInstruction>, Optional<RadixAddress>> extractInstructionsAndAuthor(byte universeMagic, Atom atom) {
+		return tuple(toInstructions(atom.getInstructions()), extractCreator(atom, universeMagic));
+	}
+
+	private static Optional<String> extractMessage(List<REInstruction> instructions) {
+		return instructions.stream()
+			.filter(instruction -> instruction.getMicroOp() == REInstruction.REOp.MSG)
+			.map(instruction -> new String(instruction.getData(), StandardCharsets.UTF_8))
```
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/ClientApiUtils.java:104]: In next PR there will be no conversion, just encoding into hex.

File: radixdlt-core/radixdlt/src/test/java/org/radix/api/jsonrpc/HighLevelApiHandlerTest.java (hunk 2 above)
Comment (siy) [reply radixdlt-core/radixdlt/src/test/java/org/radix/api/jsonrpc/HighLevelApiHandlerTest.java:146]: updated

File: radixdlt-java-common/src/main/java/com/radixdlt/crypto/ECDSASignature.java [hunk 4]
```diff
@@ -45,22 +47,40 @@
...
+	private final BigInteger r;
+	private final BigInteger s;
+	private final byte v;
 
-	public ECDSASignature() {
-		this(BigInteger.ZERO, BigInteger.ZERO);
+	private static final ECDSASignature ZERO_SIGNATURE = new ECDSASignature(BigInteger.ZERO, BigInteger.ZERO, 0);
+
+	private ECDSASignature(BigInteger r, BigInteger s, int v) {
+    	this.r = Objects.requireNonNull(r);
```
Comment (talekhinezh) [inline radixdlt-java-common/src/main/java/com/radixdlt/crypto/ECDSASignature.java:57]: spacing

**Review by talekhinezh — CHANGES_REQUESTED** [review]

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/berkeley/BerkeleyClientApiStore.java [hunk 5]
```diff
@@ -209,24 +237,87 @@ public void storeCollectedParticles() {
...
+			var list = new ArrayList<TxHistoryEntry>();
+			var count = new AtomicInteger(0);
+
+			do {
+				AID.fromBytes(data.getData())
+					.flatMap(this::retrieveTx)
+					.onFailureDo(
+						() -> log.error("Error deserializing TxID while scanning DB for address {}", address)
+					)
+					.flatMap(txn -> ClientApiUtils.toParsedTx(txn, universeMagic, sid -> store.loadUpParticle(null, sid)))
```
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/berkeley/BerkeleyClientApiStore.java:279]: I think we need to discuss it in details and find a convenient approach which will work for both cases - normal processing and restoring DB from transaction log.

File: radixdlt-core/radixdlt/src/test/java/com/radixdlt/client/store/berkeley/TransactionParserTest.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+//		executeAndDecode(List.of(ActionType.TRANSFER), UInt256.THREE, atom4);
+	}
+
+	private void executeAndDecode(
+		List<ActionType> expectedActions, UInt256 fee, Txn... txns
+	) throws RadixEngineException, InterruptedException {
+		var list = engine.execute(List.of(txns), null, PermissionLevel.USER);
+
+		// Wait for propagation
+		Thread.sleep(100L);
```
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/test/java/com/radixdlt/client/store/berkeley/TransactionParserTest.java:149]: Removed

File: radixdlt-java-common/src/main/java/com/radixdlt/crypto/ECDSASignature.java (hunk 4 above)
Comment (siy) [reply radixdlt-java-common/src/main/java/com/radixdlt/crypto/ECDSASignature.java]: fixed

File: radixdlt-core/radixdlt/src/test/java/com/radixdlt/client/store/berkeley/TransactionParserTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/test/java/com/radixdlt/client/store/berkeley/TransactionParserTest.java:145]: removed

File: radixdlt-core/radixdlt/src/test/java/com/radixdlt/client/store/berkeley/TransactionParserTest.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+		executeAndDecode(List.of(ActionType.STAKE), UInt256.ZERO, atom2);
+
+		var atom3 = TxBuilder.newBuilder(tokenOwnerAddress, store)
+			.unstakeFrom(tokenRri, validatorAddress, UInt256.ONE)
+			.signAndBuild(tokenOwnerKeyPair::sign);
+
+		executeAndDecode(List.of(ActionType.UNSTAKE), UInt256.ZERO, atom3);
+
+		//TODO: temporarily disabled because of unexpected "Could not find token rri ...." error
+//		var atom4 = TxBuilder.newBuilder(tokenOwnerAddress, store)
```
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/test/java/com/radixdlt/client/store/berkeley/TransactionParserTest.java]: I've reworked test completely and separated each case. Transfer still fails with "Could not find token RRI".

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/berkeley/BerkeleyClientApiStore.java [hunk 8]
```diff
@@ -272,62 +380,97 @@ private void addTokenWriteBytes(DatabaseEntry data) {
...
+				.map(particle -> ParticleWithSpin.create(particle, instruction.getNextSpin()));
+		} else if (instruction.getMicroOp() == REInstruction.REOp.VDOWN) {
+			return restore(serialization, instruction.getData(), Particle.class)
+				.map(particle -> ParticleWithSpin.create(particle, instruction.getNextSpin()));
+
+		} else if (instruction.getMicroOp() == REInstruction.REOp.DOWN) {
+			var substateId = SubstateId.fromBytes(instruction.getData());
+
+			return Result.fromOptional(
+				store.loadUpParticle(null, substateId),
```
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/berkeley/BerkeleyClientApiStore.java]: Reworked from scratch.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/berkeley/BerkeleyClientApiStore.java [hunk 9]
```diff
@@ -209,24 +237,87 @@ public void storeCollectedParticles() {
...
 
-		safeClose(executedTransactionsDatabase);
-		safeClose(balanceDatabase);
+		safeClose(transactionHistory);
+		safeClose(tokenDefinitions);
+		safeClose(addressBalances);
+		safeClose(supplyBalances);
+	}
+
+	private Instant instantFromKey(DatabaseEntry key) {
```
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/berkeley/BerkeleyClientApiStore.java:312]: Reworked to take timestamp from incoming transactions.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/berkeley/BerkeleyClientApiStore.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/berkeley/BerkeleyClientApiStore.java]: Reworked from scratch.

File: radixdlt-core/radixdlt/src/test/java/com/radixdlt/client/store/berkeley/TransactionParserTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/test/java/com/radixdlt/client/store/berkeley/TransactionParserTest.java]: Reworked test once again. With creation and minting of token in the same transaction everything works.

**Review by LukasGasior1 — APPROVED** [review]

**Review by talekhinezh — APPROVED** [review]

**Pull request merged**